### PR TITLE
Replace user edit form with Crispy Forms

### PR DESCRIPTION
**Pull request by sw00**
Small PR to get acquainted with Django and Crispy Forms.

### CODE REVIEW

**Comment by PiDelport** [comment]
Cool, thanks for starting this! :)

For the Travis failure, we use [isort](https://github.com/timothycrosley/isort) to automatically sort all our imports: this helps keep things consistent, and makes it so that we have one less thing to manually worry about.

To run isort, you can just run `isort --apply` in your working copy before making a commit: that will sort everything in isort's standardised order and format. (I did this in commit 6112065.) If you can, you can set up your editor to do this automatically, or with a shortcut.

**Comment by PiDelport** [comment]
The next thing we should probably do is to migrate the `buza.views.edit` view to a generic Django `UpdateView`, so that we can use our `CrispyFormMixin` to handle the common parts of adding the `FormHelper`, and customise it there.

The `QuestionCreate` / `QuestionUpdate` views (with their common code in `QuestionModelFormMixin`) are an existing example of how to approach that.

**Review by sewagodimo — DISMISSED** [review]
:cake: :beer:

**Comment by sw00** [comment]
Re https://github.com/buza-project/buza-website/pull/55/commits/2e6c0959e4369a841997591d083a4aae5f76015b: just did a haphazard change in the interest of exploring. Tests are failing but need to determine whether they still make sense with regards to new implementation or just [change detectors](https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html).

Also, sorry for the slow progress. Haven't had much time this week and just getting re-acquainted with stack after years of being away from Django/Python webdev.

File: src/buza/views.py [hunk 1]
```diff
@@ -81,6 +81,16 @@ def register(request: HttpRequest) -> HttpResponse:
     )
 
 
+class EditUser(LoginRequiredMixin, generic.UpdateView):
```
Comment (PiDelport) [inline src/buza/views.py]:
```suggestion
class UserUpdate(CrispyFormMixin, LoginRequiredMixin, generic.UpdateView):
```

File: src/buza/urls.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
     # user related paths
     url(r'^register/$', views.register, name='register'),
-    url(r'^edit/$', views.edit, name='edit'),
+    path(r'edit/<int:pk>/', views.EditUser.as_view(), name='edit'),
```
Comment (PiDelport) [inline src/buza/urls.py]:
```suggestion
    path(r'edit/<int:pk>/', views.UserUpdate.as_view(), name='user-update'),
```

File: src/buza/views.py [hunk 3]
```diff
@@ -81,6 +81,16 @@ def register(request: HttpRequest) -> HttpResponse:
...
+    model = models.User
+    fields = ['phone', 'school', 'school_address', 'grade', 'photo', 'bio', 'subjects']
+
+    def get_form_helper(self, form: forms.ModelForm) -> FormHelper:
+        helper = super().get_form_helper(form)
+        helper.form_action = reverse('edit')
```
Comment (PiDelport) [inline src/buza/views.py]:
```suggestion
        helper.form_action = reverse('user-update')
```

**Comment by PiDelport** [comment]
@sw00: Cool! The tests are failing because the view's URL now actually takes a parameter: they should be updated for that.

We'll also need to add some authorisation, so that users can only edit themselves (as an initial policy), for which we can probably pull in https://github.com/pjdelport/django-auth-utils

**Comment by PiDelport** [comment]
@sw00: Okay, I added some changes, removed the old edit view and form that this replaces, and update the tests. How does this look to you? (And thanks for getting this started! :D)

@sewagodimo: The actual design could probably use more tweaking as we go on, but this PR should be good to merge as it stands. Are you happy with it?

File: src/buza/views.py [hunk 4]
```diff
@@ -81,30 +78,43 @@ def register(request: HttpRequest) -> HttpResponse:
...
+    # TODO: Merge with and migrate to existing buza/user_form.html ?
+    template_name = 'accounts/edit.html'
+
+    def get_object(self, queryset: QuerySet = None) -> models.User:
+        """
+        Only allow users to update their own profile.
+        """
+        user: models.User = super().get_object(queryset)
+        if not user == self.request.user:
+            raise PermissionDenied('You can only update your own profile.')
```
Comment (PiDelport) [inline src/buza/views.py:104]: We should ideally test this check before merging…

**Review by PiDelport — CHANGES_REQUESTED** [review]

**Review by PiDelport — APPROVED** [review]

**Comment by PiDelport** [comment]
Let's get this merged! Thanks. :coffee:

**Review by sewagodimo — COMMENTED** [review]
approved

**Pull request merged**